python/lsst/daf/butler/core/datastoreCacheManager.py

```diff
@@ -223,21 +497,277 @@ def move_to_cache(self, uri: ButlerURI, ref: DatasetRef) -> Optional[ButlerURI]:
+            If `True` the item being registered can already be listed.
+            This can allow a cache refresh to run without checking the
+            file again. If `False` it is an error for the registry to
+            already know about this file.
+
+        Returns
+        -------
+        cache_key : `str`
+            The key used in the registry for this file.
+        """
+        path_in_cache = cached_location.relative_to(self.cache_directory)
+        if path_in_cache is None:
+            raise ValueError(f"Can not register cached file {cached_location} that is not within"
+                             f" the cache directory at {self.cache_directory}.")
+        if path_in_cache in self._cache_entries:
+            if can_exist:
+                return path_in_cache
+            else:
+                raise ValueError(f"Cached file {cached_location} is already known to the registry"
+                                 " but this was expected to be a new file.")
+        details = CacheEntry.from_file(cached_location, root=self.cache_directory)
+        self._cache_entries[path_in_cache] = details
+        return path_in_cache
+
+    def scan_cache(self) -> None:
+        """Scan the cache directory and record information about files.
+        """
+        found = set()
+        for file in ButlerURI.findFileResources([self.cache_directory]):
+            assert isinstance(file, ButlerURI), "Unexpectedly did not get ButlerURI from iterator"
+
+            # Skip any that are found in an exempt part of the hierarchy
+            # since they should not be part of the registry.
+            if file.relative_to(self._temp_exempt_directory) is not None:
+                continue
+
+            path_in_cache = self._register_cache_entry(file, can_exist=True)
+            found.add(path_in_cache)
+
+        # Find any files that were recorded in the cache but are no longer
+        # on disk. (something else cleared them out?)
+        known_to_cache = set(self._cache_entries)
+        missing = known_to_cache - found
+
+        if missing:
+            log.debug("Entries no longer on disk but thought to be in cache and so removed: %s",
+                      ",".join(missing))
+            for path_in_cache in missing:
+                self._cache_entries.pop(path_in_cache)
+
+    def _remove_from_cache(self, cache_entries: Iterable[str]) -> None:
+        """Remove the specified cache entries from cache.
+
+        Parameters
+        ----------
+        cache_entries : iterable of `str`
+            The entries to remove from the cache. The values are the path
+            within the cache.
+        """
+        for entry in cache_entries:
+            path = self.cache_directory.join(entry)
+
+            # If the file happens to also exist in the exempt folder
+            # skip removal since a process is actively using the file.
+            if self._temp_exempt_directory.join(entry).exists():
```

Comment:
I'm worried that there still may be (brief but non-zero) windows where a race can happen.  Still, since you treat it as uncached in that case, it should be safe if less performant.

Reply:
I did think about using hard links exclusively and letting the cache expiration delete the file regardless. Then I worried that someone might put the cache dir on some file system where hard links wouldn't work for some reason and so I played it a bit safer by skipping the removal if it existed in both places. I'm not sure how to deal with the case of another process deleting the file in the instant before I create a soft link in the exempt directory (the hardlink race would cause the hardlink to fail to be made but would helpfully fall back to creating a soft link to a file that is no longer there).

Reply:
Do you want to block merging until we try to make this robust? How much can we rely on hardlinks working within the temp directory? Hardlinks working mean that we can remove this specific test completely and try to make a hardlink. If the hardlink is made then there is no race (deleting the cache file won't delete the thing in the "exempt" directory). If the hardlink fails we will have to report that the file wasn't really cached. That is all great assuming hardlinks can be trusted. Softlinks are the reason there is a possible race.

Reply:
hardlinks are much cleaner because then the cache directory can expire as normal without having to keep the extra files around and the "exempt" directory can delete things independently (let's hope they aren't kept around for too long if disk space is tight)

Reply:
Since these are on the worker and all within the same directory tree, I think hardlinks are reliable.  But having any race be "uncached" is also a belt-and-suspenders safety precaution.

Reply:
I've removed the softlink option which also removes any need for cache expiry to know about the "exempt" directory. The cache doesn't have to be on the worker -- people can specify a directory that is on some shared file system if they want.